### PR TITLE
Segregated Witness support

### PR DESCRIPTION
**Pull request by mcelrath**
This is my initial implementation.  It correctly serializes and deserializes P2WPKH segwit transactions.  I haven't tested the other transaction types.

Please don't merge this yet, but I wanted to make it available for others to work on.

TODO
- ~~BIP 143 tx signing~~
- ~~`CBlock` modifications to support witness data~~
- ~~P2WSH and BIP16 embedded scripts~~
- ~~tests!~~

### CODE REVIEW

File: bitcoin/core/__init__.py [hunk 1]
```diff
@@ -336,12 +339,50 @@ def from_txinwitness(cls, txinwitness):
...
+        object.__setattr__(self, 'vtxinwit', vtxinwit)
+
+    def is_null(self):
+        for n in range(len(self.vtxinwit)):
+            if not self.vtxinwit[n].is_null(): return False
+        return True
+
+    # FIXME this cannot be a @classmethod like the others because we need to
+    # know how many items to deserialize, which comes from len(vin)
+    def stream_deserialize(self, f):
```
Comment (petertodd) [inline bitcoin/core/__init__.py]: How about we add an expected_length parameter instead?

File: bitcoin/core/__init__.py [hunk 2]
```diff
@@ -422,11 +461,9 @@ def GetTxid(self):
...
             given by GetHash.  GetTxid excludes witness data, while GetHash
             includes it. """
-        if self.wit:
-            wit = self.wit
-            self.wit = b''
-            txid = Hash(self.serialize())
-            self.wit = wit
+        if self.wit != CTxWitness():
+            txid = Hash(CTransaction(self.vin, self.vout, self.nLockTime,
+                self.nVersion).serialize())
```
Comment (petertodd) [inline bitcoin/core/__init__.py]: Seems pretty dangerous to me to change the meaning of GetHash() on everyone - Bitcoin Core's codebase gives GetHash() the same behavior as before.

How about we remove GetHash() entirely and replace it with GetTxid() and GetWitnessHash()? Better for code to cleanly break than do the wrong thing.

**Comment by afk11** [comment]
This PR contains much of the ground work for segwit, perhaps it'd be good to get this merged so other areas can be addressed.

**Comment by sbaks0820** [comment]
Does this PR also implement the deserialization changes necessary to use RPC with segwit transactions? I'm working on testnet and segwit transactions are already showing up there

**Comment by mcelrath** [comment]
Yes, it correctly deserializes segwit transactions on testnet. I've been using it for this for months. 

The only thing missing is deserialization of blocks containing a witness block. 

@afk11 I'll try to get this merged the first week of January when I'm back from holiday.

On December 25, 2016 5:49:33 PM CST, Surya Bakshi <notifications@github.com> wrote:
>Does this PR also implement the deserialization changes necessary to
>use RPC with segwit transactions? I'm working on testnet and segwit
>transactions are already showing up there
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/petertodd/python-bitcoinlib/pull/112#issuecomment-269142187
>
>!DSPAM:58605b15305241940418538!

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by sbaks0820** [comment]
Possible bug when going a `SignatureHash`. The function executes `script.is_witness_scriptpubkey()`, line [1018](https://github.com/VidaID/python-bitcoinlib/blob/master/bitcoin/core/script.py#L1018) / [680](https://github.com/VidaID/python-bitcoinlib/blob/master/bitcoin/core/script.py#L680),  which checks `CScriptOp(self[0]).is_small_int()`. `CScriptOp(self[0])` fails because `self[0]` is interpreted as a `str` instead of an `int` so when `__new__` runs `return _opcode_instances[n]`, line [106](https://github.com/VidaID/python-bitcoinlib/blob/master/bitcoin/core/script.py#L106), it throws `TypeError: list indices must be integers, not str`.

Reproduce
---------------
run the `spend-p2pkh-txout.py` example in python 2, the example should throw the error on `SignatureHash(...)`

Simple fix is probably just to cast `self[0]` to an `int`.

**Comment by amiller** [comment]
Is there any code for *verifying* a SegWit transaction? It seems like the hard part (computing the sighash especially) is already there, so this should just require adding a variant of the `VerifyScript` function.

**Comment by afk11** [comment]
@amiller I don't think there is yet, though segwit verification is actually a light enough task. Mostly it's just writing VerifyWitnessProgram and modifying VerifyScript to call it. 

Similar amount of work might be using FFI to call functions exposed by bitcoinconsensus.h - the shared library is included in gitian builds, and since 0.13.0 has had segwit verification support through `bitcoinconsensus_verify_script_with_amount`

**Comment by unsystemizer** [comment]
I tried this patch but I still get the same error (on testnet; works fine with mainnet as others have mentioned). @sbaks0820, after those changes (change self[0] to struct.unpack('<b', self[0])) can you work on testnet?

**Comment by sbaks0820** [comment]
Yes, the solution works with both testnet and mainnet. If you use the same `spend-p2pkh-txout.py` example but change the parameters, txid and recipient address to testnet hashes, the solution works.

**Comment by afk11** [comment]
Ping @mcelrath @petertodd - think transaction signing is enough for this PR to be merged? (perhaps waiting for tests?) It's likely most people interested in using segwit have no interest in the witness commitment part, it could come as another pull request.

**Comment by mcelrath** [comment]
Thanks to @the9ull for block serialization code and tests, and @sbaks0820 for a bugfix.

@petertodd I'd say this is ready to be merged.  The only outstanding issue might be your comments above about an `expected_length` parameter?

**Comment by sbaks0820** [comment]
Anyone have an example of how to create a seg wit transaction from this fork that spends a P2PKH output? Is seg wit limited to P2PSH?

**Comment by robby-dermody** [comment]
Just checking in on when we can see this merged into `master`. Thanks!

**Comment by petertodd** [comment]
Sorry for taking so long; no-one is paying me to do Bitcoin-Core-related work right now, and I have a backlog of non-Core work that I'm trying to get through. Be a few more weeks at least before I'll be able to spend a few days on this. :(

**Comment by unsystemizer** [comment]
Any chance to get this merged before SW gets activated on mainnet?

**Comment by deweller** [comment]
Hey Peter.  Can you review and merge this soon?  We would like to use this library to support segwit.  Thanks.

File: bitcoin/core/script.py [hunk 3]
```diff
@@ -680,7 +680,7 @@ def is_p2sh(self):
     def is_witness_scriptpubkey(self):
         """Returns true if this is a scriptpubkey signaling segregated witness
         data. """
-        return 3 <= len(self) <= 42 and CScriptOp(self[0]).is_small_int()
```
Comment (zzir) [inline bitcoin/core/script.py]: I running successfully on this line.

**Comment by petertodd** [comment]
Merged!

Note that I may still tweek the API further - I want to try using this new support in my own programs, and see how others use it, before deciding on exactly how the API should work. So if you choose to use this in a project, I'd strongly suggest using git subtree/submodule to reference *exactly* what git commit you're using in your library.

**Comment by rubensayshi** [comment]
any plan for a tagged release so we can start using it?

**Pull request merged**